File: CSharpCombined/Client/FormCSharp/ClientListener.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace ClientSolution
{
    class ClientListener
    {
        private Client client;

        public ClientListener(Client client) { this.client = client; }

        public void Run()
        {
            StreamReader reader = client.reader;

            while (true)
            {
                try
                {
                    string line = reader.ReadLine();
                    if(line == null)
                    {
                        line = "";
                    }

                    string[] substrings = line.Split(' ');

                    string command = substrings[0];
                    string firstParameter = substrings[1];

                    switch (command.ToUpper())
                    {
                        case "ASSIGN_ID":
                            int id = int.Parse(firstParameter);
                            client.playerID = id;
                            break;

                        case "SEND_GAMESTATE":
                            string players = firstParameter;
                            string playerWithBall = substrings[2];
                            Console.WriteLine("Got gamestate");
                            client.SetGamestate(players, playerWithBall);
                            break;
                        case "MOVE_BALL":
                            client.NewBallPosition(int.Parse(firstParameter));
                            break;

                        case "PLAYER_JOIN":
                            client.AddNewPlayer(firstParameter);
                            break;

                        case "PLAYER_LEAVE":
                            client.RemovePlayer(firstParameter);
                            break;

                        case "ERROR":
                            client.ShowErrorToClient(line);
                            break;
                    }
                }
                catch(Exception)
                {
                    Console.WriteLine("Server disconnected");
                    Console.ReadLine();
                    break;
                }
            }

        }
    }
}

File: JavaCombined/src/Client/ClientListener.java
package Client;

import java.io.IOException;
import java.util.NoSuchElementException;
import java.util.Scanner;

public class ClientListener implements Runnable {

    private Client client;
    private boolean autoRestart = true;

    public ClientListener(Client client) {
        this.client = client;
    }

    //constructor for testing -- allows disable of auto-restart
    public ClientListener(Client client, boolean autoRestart) {
        this.client = client;
        this.autoRestart = autoRestart;
    }

    @Override
    public void run() {
        Scanner reader = client.getReader();

        while (true) {
            try {
                String line = reader.nextLine();
                String[] substrings = line.split(" ");
                String command = substrings[0];
                String firstParameter = substrings[1];

                switch (command.toUpperCase()) {
                    case "SEND_GAMESTATE":
                        String players = substrings[1];
                        String playerWithBall = substrings[2];
                        client.setGamestate(players, playerWithBall);
                        break;
                    case "MOVE_BALL":
                        client.newBallPosition(Integer.parseInt(firstParameter));
                        break;
                    case "PLAYER_JOIN":
                        client.addNewPlayer(firstParameter);
                        break;
                    case "PLAYER_LEAVE":
                        client.removePlayer(firstParameter);
                        break;
                    case "ERROR":
                        client.showErrorToClient(line);
                        break;
                }
            }
            //Thrown when connection is lost
            catch (NoSuchElementException e) {
                System.out.println("Server disconnected.");
                if (autoRestart) {
                    try {
                        client.startNewServer();
                    } catch (IOException ex) {
                        System.out.println("Failed to start new server.");
                        ex.printStackTrace();
                    }
                }
                return;
            }
        }
    }
}
